triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Server.scala

```diff
@@ -171,20 +188,102 @@ class Server(
+                inner(Tuple1(Some(auth.accessToken)))(ctx)
+              }
+            case StatusCodes.Unauthorized =>
+              val requestId = UUID.randomUUID
+              authRequests.update(requestId, ctx.request)
+              val uri = authUri
+                .withPath(Path./("login"))
+                .withQuery(
+                  AuthRequest
+                    .Login(
+                      // TODO[AH] Make the redirect URI configurable, especially the authority. E.g. when running behind nginx.
+                      Uri()
+                        .withScheme(ctx.request.uri.scheme)
+                        .withAuthority(ctx.request.uri.authority)
+                        .withPath(Path./("cb")),
+                      claims,
+                      Some(requestId.toString),
+                    )
+                    .toQuery)
+              ctx.redirect(uri, StatusCodes.Found)
+            case statusCode =>
+              Unmarshal(resp).to[String].flatMap { msg =>
+                logger.error(s"Failed to authorize with middleware ($statusCode): $msg")
+                ctx.complete(
+                  errorResponse(
+                    StatusCodes.InternalServerError,
+                    "Failed to authorize with middleware"))
+              }
+          }
+        } yield result
+    }
+  }
+
+  private def authCallback(
+      requestId: UUID)(implicit ec: ExecutionContext, system: ActorSystem): Route = ctx => {
+    authRequests.remove(requestId) match {
+      case None => complete(StatusCodes.NotFound)(ctx)
+      case Some(req) =>
+        // Replay the stored request forwarding any new cookies.
+        val newRequest = req
+          .addHeader(headers.Cookie(cookies = ctx.request.cookies))
+        val newCtx = ctx
+          .withRequest(newRequest)
+          .withUnmatchedPath(newRequest.uri.path)
+        route(ec, system)(newCtx)
+    }
+  }
+
+  private def route(implicit ec: ExecutionContext, system: ActorSystem) = concat(
     post {
       concat(
         // Start a new trigger given its identifier and the party it
         // should be running as.  Returns a UUID for the newly
         // started trigger.
         path("v1" / "start") {
-          entity(as[StartParams]) { params =>
-            startTrigger(params.party, params.triggerName) match {
-              case Left(err) =>
-                complete(errorResponse(StatusCodes.UnprocessableEntity, err))
-              case Right(triggerInstance) =>
-                complete(successResponse(triggerInstance))
-            }
+          entity(as[StartParams]) {
+            params =>
+              val claims =
+                AuthRequest.Claims(actAs = List(Ref.Party.assertFromString(params.party.toString)))
+              // TODO[AH] Why do we need to pass ec, system explicitly?
+              authorize(claims)(ec, system) { token =>
+                startTrigger(params.party, params.triggerName, token) match {
```

Comment:
We obviously also need this on other endpoints but I assume we’re going to add this later.

Reply:
I've started with this one since it is the one where the token is actually required for the interaction with the ledger. AFAIK the others don't require authorization as far as the ledger is concerned. Do we also want to use the auth middleware to double as an authorization service for actions on the trigger service itself? E.g. uploading a DAR or stopping a trigger.

Reply:
Thanks for clarifying. Stop seems straight forward enough, upload sounds like it needs a bit more thinking. I'll leave these for future PRs.